File: packages/cli-py/whale_tutor/runner.py
"""subprocess 启动 node server,信号转发(Ctrl+C 优雅退出)。"""

from __future__ import annotations

import os
import shutil
import signal
import socket
import subprocess
import sys
import threading
import time
import webbrowser
from pathlib import Path

from rich.console import Console

from .config import WhaleTutorConfig, to_env

console = Console()


def find_node() -> str:
    """找系统中的 node。找不到给清晰的错误提示。"""
    node = shutil.which("node")
    if not node:
        console.print(
            "[red]✗ 未找到 node。Whale Tutor 需要 Node.js ≥ 22。"
            "请装 https://nodejs.org/ 后重试。[/red]"
        )
        sys.exit(1)
    return node


def _wait_and_open_browser(port: int, timeout_s: float = 30.0) -> None:
    """轮询 :port 直到能 connect 上,然后开浏览器。在线程里跑。"""
    deadline = time.time() + timeout_s
    url = f"http://localhost:{port}"
    while time.time() < deadline:
        try:
            with socket.create_connection(("127.0.0.1", port), timeout=0.5):
                # 给 nest 的 router 再一点点时间(server listen != routes ready)
                time.sleep(0.3)
                webbrowser.open(url)
                return
        except OSError:
            time.sleep(0.3)
    # 超时不报错 — server 可能有别的问题,日志会显示


def start_node_server(
    cfg: WhaleTutorConfig, bundle_root: Path, *, open_browser: bool = True
) -> int:
    """启动 node server,前台运行直到 server 退出或用户 Ctrl+C。

    返回 node 进程退出码。open_browser=True 时 server ready 后自动开默认浏览器。
    """
    node = find_node()
    server_main = bundle_root / "server" / "dist" / "main.js"
    web_dir = bundle_root / "web"
    schema_file = bundle_root / "db" / "init" / "01-schema.sql"

    if not server_main.exists():
        console.print(
            f"[red]✗ 找不到 server bundle ({server_main})。"
            "可能是 pip 包损坏,试试重装 whale-tutor。[/red]"
        )
        sys.exit(1)

    env = os.environ.copy()
    env.update(to_env(cfg, web_dir=web_dir, schema_file=schema_file))

    console.print(
        f"[bold green]→ 启动 server[/bold green] [dim]node {server_main.name}[/dim]"
    )
    console.print(f"[dim]  courses: {cfg.courses_dir}[/dim]")
    console.print(f"[dim]  port: {cfg.server.port}[/dim]")
    console.print(
        f"[dim]  db: {cfg.database.user}@{cfg.database.host}:"
        f"{cfg.database.port}/{cfg.database.database}[/dim]"
    )
    console.print(
        f"[bold]  → 浏览器打开 http://localhost:{cfg.server.port}[/bold]\n"
    )

    # 用 Popen 而非 run,这样可以在主进程接 SIGINT 后转发给 child
    # cwd 设为 server bundle 内,让 nest-cli.json 等相对路径生效
    proc = subprocess.Popen(
        [node, str(server_main)],
        env=env,
        cwd=str(server_main.parent),
    )

    if open_browser:
        threading.Thread(
            target=_wait_and_open_browser,
            args=(cfg.server.port,),
            daemon=True,
        ).start()

    def forward_signal(signum, _frame):  # noqa: ANN001
        # 让 node 收到信号,优雅退出
        try:
            proc.send_signal(signum)
        except ProcessLookupError:
            pass

    # SIGINT (Ctrl+C) 在 Windows 上语义不同,但 Popen.send_signal 会用平台原生方式
    signal.signal(signal.SIGINT, forward_signal)
    if hasattr(signal, "SIGTERM"):
        signal.signal(signal.SIGTERM, forward_signal)

    try:
        return proc.wait()
    except KeyboardInterrupt:
        # 兜底:Python KeyboardInterrupt
        try:
            proc.terminate()
            return proc.wait(timeout=5)
        except subprocess.TimeoutExpired:
            proc.kill()
            return proc.wait()

File: packages/cli-node/lib/runner.mjs
// spawn node server + 信号转发(Ctrl+C 优雅退出) + ready 后自动开浏览器。

import { spawn } from 'node:child_process';
import { existsSync } from 'node:fs';
import net from 'node:net';
import { dirname, join } from 'node:path';
import process from 'node:process';
import kleur from 'kleur';
import { toEnv } from './config.mjs';

// 轮询 :port 直到能 connect 上,然后(若 openBrowser)用平台默认浏览器打开。
// 探测成功才打印 "✓ Server 就绪" — 之前用户不该访问 URL(会 connection refused)。
async function waitForReady(port, { openBrowser, timeoutMs = 30000 }) {
  const deadline = Date.now() + timeoutMs;
  const probe = () =>
    new Promise((resolveProbe) => {
      const sock = net.createConnection({ host: '127.0.0.1', port }, () => {
        sock.end();
        resolveProbe(true);
      });
      sock.on('error', () => resolveProbe(false));
      sock.setTimeout(500, () => {
        sock.destroy();
        resolveProbe(false);
      });
    });

  while (Date.now() < deadline) {
    if (await probe()) {
      // 给 nest 路由再一点点时间(server listen != routes ready)
      await new Promise((r) => setTimeout(r, 300));
      const url = `http://localhost:${port}`;
      console.log('\n' + kleur.green().bold(`✓ Server 就绪`) + kleur.dim(`  → ${url}`));
      if (openBrowser) {
        console.log(kleur.dim(`  浏览器自动打开中…(没弹出请手工访问上面 URL)`));
        const opener =
          process.platform === 'win32'
            ? ['cmd', ['/c', 'start', '""', url]]
            : process.platform === 'darwin'
              ? ['open', [url]]
              : ['xdg-open', [url]];
        try {
          spawn(opener[0], opener[1], { detached: true, stdio: 'ignore' }).unref();
        } catch {
          // 没装 xdg-open 等场景:静默放弃,日志里有 url
        }
      } else {
        console.log(kleur.dim(`  --no-open 已设,请手工访问上面 URL`));
      }
      return;
    }
    await new Promise((r) => setTimeout(r, 300));
  }
  // 探测超时 — server 可能没起来或卡在初始化(连不上 mysql / 找不到 web bundle 等)
  console.log(
    '\n' +
      kleur.yellow(
        `⚠ 等了 ${timeoutMs / 1000} 秒 server 仍未在 :${port} listen,看上方日志找原因(常见:mysql 连不上 / 端口被占)。`,
      ),
  );
}

// 启动 node server,前台运行直到 server 退出或用户 Ctrl+C。
// 返回 child 进程 exit code。openBrowser=true 时 ready 后自动打开默认浏览器。
export async function startServer(cfg, bundleRoot, { openBrowser = true } = {}) {
  const serverMain = join(bundleRoot, 'server', 'dist', 'main.js');
  const webDir = join(bundleRoot, 'web');
  const schemaFile = join(bundleRoot, 'db', 'init', '01-schema.sql');

  if (!existsSync(serverMain)) {
    console.error(
      kleur.red(`✗ 找不到 server bundle (${serverMain})。可能是 npm 包损坏,试试重装 whale-tutor。`),
    );
    process.exit(1);
  }

  console.log(kleur.green().bold(`→ 启动 server  `) + kleur.dim('node main.js'));
  console.log(kleur.dim(`  courses: ${cfg.coursesDir}`));
  console.log(kleur.dim(`  port: ${cfg.server.port}`));
  console.log(
    kleur.dim(
      `  db: ${cfg.database.user}@${cfg.database.host}:${cfg.database.port}/${cfg.database.database}`,
    ),
  );
  console.log(
    kleur.yellow(
      `  ⏳ 等 server 启动中,看到下方 "✓ Server 就绪" 再访问 http://localhost:${cfg.server.port}\n`,
    ),
  );

  const child = spawn(process.execPath, [serverMain], {
    env: { ...process.env, ...toEnv(cfg, { webDir, schemaFile }) },
    cwd: dirname(serverMain),
    stdio: 'inherit',
  });

  // 不论是否 auto-open browser,都轮询端口 — 让用户知道 server 真正 ready 的时刻
  waitForReady(cfg.server.port, { openBrowser }).catch(() => {
    // 不阻塞,不污染输出
  });

  // 信号转发 — Ctrl+C 让 child 优雅退出
  const forward = (sig) => () => {
    if (!child.killed) child.kill(sig);
  };
  process.on('SIGINT', forward('SIGINT'));
  process.on('SIGTERM', forward('SIGTERM'));

  return new Promise((resolveExit) => {
    child.on('exit', (code) => resolveExit(code ?? 0));
  });
}
